### DISCUSSION TITLE
ll-hls does not working with native player IOS/Safari

### DISCUSSION THREAD

**Discussion by driwn** [OvenMediaLabs/OvenMediaEngine#1253] [Help] [answered]
Trying to resolve this issue https://github.com/AirenSoft/OvenPlayer/issues/357

ll-hls does not working with native player IOS/Safari (trying IPhone 13 pro max, IOS 16.4.1)

Working well with ovenplayer on windows, but getting error on iphone,

through ovenplayer or video tag the same result

segments request cancelled immediately or after first chunk

input (h264, aac) as far as I know
I don't know more details

the screenshot shows codecs (avc1.42c01f mp4a.40.2)

but the base hls with other engines worked well

cameras => ATEM mini pro => ATEM software => rtmp => OME => ll-hls

**Comment by getroot** [comment]
I tested using your Server.xml as is, but it plays fine in iOS Safari. I used OvenMediaEngine 0.15.11 and the dev version. Encoder was OBS.

What version of OME are you using? Does it reproduce using OBS instead of ATEM mini pro?

**Reply by driwn** [reply]
OME 0.15.9
idk about OBS now, but maybe I can just encode the audio (or video too) inside the OME to the necessary codecs? which ones need?

**Reply by getroot** [reply]
First of all, it would be good to check if the problem is reproduced with OBS.

I don't understand your question very well, maybe what you want is the link below?
https://airensoft.gitbook.io/ovenmediaengine/transcoding

**Reply by driwn** [reply]
Doesn't work with OBS either.

**Reply by getroot** [reply]
This problem is not reproducible in any environment I manage.

Does the URL below play fine on your iOS?
https://llhls-demo.ovenmediaengine.com/app/stream/llhls.m3u8

I can analyze further if you share your Server.xml and OvenMediaEngine.log files and your LLHLS URL.

**Reply by driwn** [reply]
this url is working fine, the rest I emailed

**Reply by getroot** [reply, marked as answer]
Your server doesn't seem to be fast enough to handle your setup. This could be your network speed issue, or your CPU performance issue. Or it could be that the input stream from the encoder is coming in slowly. That is, chunk generation is slow, or the network speed is slow. In this case, Safari gives up playback.
Please use the latest version of OvenMediaEngine first.

```
<LLHLS>
        <ChunkDuration>0.2</ChunkDuration>
        <SegmentDuration>6</SegmentDuration>
        <SegmentCount>4</SegmentCount>
        <CrossDomains>
                <Url>*</Url>
        </CrossDomains>
</LLHLS>
```

A ChunkDuration of 0.2 means that the player should be able to download a partial segment in less than 0.2 seconds on average. But as you can see from the capture, your server doesn't. Also, since SegmentCount is 4, Safari also fails to switch to legacy HLS.

![image](https://github.com/AirenSoft/OvenMediaEngine/assets/17779085/1328e27a-7c30-453b-9ad0-d4ba5fe7c2bc)


Could you try with the settings below?

```
<LLHLS>
	<ChunkDuration>1</ChunkDuration>
	<PartHoldBack>3.0</PartHoldBack>
	<SegmentDuration>6</SegmentDuration>
	<SegmentCount>10</SegmentCount>
	<CrossDomains>
		<Url>*</Url>
	</CrossDomains>
</LLHLS>
```

**Reply by driwn** [reply]
it works with this settings, but latency increases from 3s to ~15s, 
CPU is OK, network is slow? maybe I can use another optimal settings?

also sometimes stream drops down, maybe I can add reload in ovenplayer?

**Reply by getroot** [reply]
As mentioned above, check the CPU, Network speed, and input stream speed of the encoder comprehensively.

**Reply by naanlizard** [reply]
> Could you try with the settings below?
> 
> ```
> <LLHLS>
> 	<ChunkDuration>1</ChunkDuration>
> 	<PartHoldBack>3.0</PartHoldBack>
> 	<SegmentDuration>6</SegmentDuration>
> 	<SegmentCount>10</SegmentCount>
> 	<CrossDomains>
> 		<Url>*</Url>
> 	</CrossDomains>
> </LLHLS>
> ```

Can you clarify when we should set PartHoldBack and increase ChunkDuration?

PartHoldBack isn't listed in the docs, so I'm not sure what it does, and I am not sure about how we calculate the expected latency from the other three settings. Is .2->1 for ChunkDuration causing the increase in LLHLS latency from 3s to 15s?

I ask because we're debugging some consistent, but unpredictable stuttering with ovenplayer and I want to make sure OME settings are optimal to begin with.

**Reply by naanlizard** [reply]
And a follow up question - should we expect higher CPU usage with a smaller chunkduration?

We switched to 1s chunkduration and buffering seems to happen less, which is great, and CPU usage seems lower as well. But I'm not sure it's causal

**Reply by getroot** [reply]
@naanlizard As for PART-HOLD-BACK, reading the RFC would be better than my explanation. This is related to target latency as judged by the player. (Of course, some players don't follow this, so it can be confusing.)

https://datatracker.ietf.org/doc/html/draft-pantos-hls-rfc8216bis#section-4.4.3.8

Latency in LLHLS can be very dependent on the player's algorithm. In particular, it has the most to do with the player's buffer size. The LLHLS player tries to keep the buffer as large as possible if the network is slow. The specification of LLHLS is not to guarantee latency, but to help keep the player's buffer small. As an aside, WebRTC players tend to drop packets while keeping their buffers small.

Shorter chunk duration means higher PPS. 10 PPS (chunklist request + partial segment download request) per second when 0.2, 2 PPS per second when 1. This can of course affect CPU occupancy.

**Comment by xiaozhihong** [comment]
The same issue.

![企业微信20230705-114741](https://github.com/AirenSoft/OvenMediaEngine/assets/8449258/1d44fc4e-4a36-4ca8-a98d-812311b946cf)

This [demo stream](https://llhls-demo.ovenmediaengine.com/app/stream/llhls.m3u8) works well in Chrome + hls.js(left side in picture), but has high latency in Safari native player(right side in picture), it seems that LL-HLS not working.

**Reply by getroot** [reply]
Check the seek bar of the Safari player,
Safari may start in legacy HLS mode depending on various factors such as network speed. At this time, if you check the seek bar, you can see that it is in the very past, and if you seek recently, it operates in ll-hls mode.

**Reply by xiaozhihong** [reply]
![image](https://github.com/AirenSoft/OvenMediaEngine/assets/8449258/f3fd7ef1-fd0e-4a78-bf75-01a510df55d6)
No seek bar in live mode. :(
Can not seek or drag

**Reply by getroot** [reply]
I don't have a MAC and on my iOS it looks like this.
![image](https://github.com/AirenSoft/OvenMediaEngine/assets/17779085/a1ec0ba8-ec54-4c2e-92e7-843de07e8ce6)

I think it depends on the Safari platform or version. What version of Safari are you using?

**Reply by xiaozhihong** [reply]
My Os:  macOS Ventura 13.4.1
Brower: Safari 16.5.1 (18615.2.9.11.7)

I test another LL-HLS streams like [https://stream.mux.com/v69RSHhFelSm4701snP22dYz2jICy4E4FUyk02rW4gxRM.m3u8](https://stream.mux.com/v69RSHhFelSm4701snP22dYz2jICy4E4FUyk02rW4gxRM.m3u8
), it works well

<img width="1727" alt="image" src="https://github.com/AirenSoft/OvenMediaEngine/assets/8449258/f0b49aa1-08eb-46f0-807d-58e47dd583c0">

**Reply by getroot** [reply]
Well... In Seoul, the MUX test server runs legacy HLS on my iOS. And the MUX test server doesn't even work with seek on my iOS.
![IMG_6628](https://github.com/AirenSoft/OvenMediaEngine/assets/17779085/1a457f23-ce47-4900-b237-bbe331ecc00b)

I think the MUX test server is fast from your country and the OME test server is slow.

If you have any more information to learn about Safari, please share.

**Reply by xiaozhihong** [reply]
Okay, I will keep track of it. Thanks again.

**Comment by captain-bugs** [comment]
I too have issues with the LLHLS in ios AVPlayer safari player. AVPlayer and safari's player is far behind in playing compared to OvenPlayer. Stream works fine on hls.js on web and ExoPlayer on android.
This is my `Server.xml`

				<Application>
					<Name>app</Name>
					
					<Type>live</Type>
					
					<OutputProfiles>
						
						<OutputProfile>
							<Name>bypass_stream</Name>
							<OutputStreamName>${OriginStreamName}</OutputStreamName>

						<Playlist>
								<Name>For LLHLS</Name>
								<FileName>abr</FileName>

								<Rendition>
									<Name>Bypass</Name>
									<Video>source</Video>
									<Audio>aac_audio</Audio>
								</Rendition>
								<Rendition>
									<Name>FHD</Name>
									<Video>uhd_60</Video>
									<Audio>aac_audio</Audio>
								</Rendition>
								<Rendition>
									<Name>HD</Name>
									<Video>hi_30</Video>
									<Audio>aac_audio</Audio>
								</Rendition>
							</Playlist>


							<Encodes>
								<Video>
									<Name>source</Name>
									<Bypass>true</Bypass>
								</Video>
								
								<Audio>
									<Name>aac_audio</Name>
									<Codec>aac</Codec>
									<Bitrate>128000</Bitrate>
									<Samplerate>48000</Samplerate>
									<Channel>2</Channel>
									<BypassIfMatch>
										<Codec>eq</Codec>
									</BypassIfMatch>
								</Audio>								
															
								<Video>
									<Name>uhd_60</Name>
									<Codec>h264</Codec>
									<Bitrate>1677721</Bitrate>
									<Framerate>60</Framerate>
									<Width>1920</Width>
									<Height>1080</Height>
									<Preset>faster</Preset>
								</Video>

								<Video>
									<Name>hi_30</Name>
									<Codec>h264</Codec>
									<Bitrate>2024000</Bitrate>
									<Framerate>30</Framerate>
									<Width>1280</Width>
									<Height>720</Height>
									<Preset>faster</Preset>
								</Video>

								
							</Encodes>

						</OutputProfile>

					</OutputProfiles>

					<Providers>
						<RTMP />
						
						<Schedule>
							<MediaRootDir>/opt/ovenmediaengine/media</MediaRootDir>
							<ScheduleFilesDir>/opt/ovenmediaengine/media</ScheduleFilesDir>
						</Schedule>
						
					</Providers>
					
					<Publishers>
						<AppWorkerCount>1</AppWorkerCount>
						<StreamWorkerCount>8</StreamWorkerCount>
						
						<LLHLS>

							<OriginMode>true</OriginMode>
							
							<CacheControl>
								<MasterPlaylistMaxAge>0</MasterPlaylistMaxAge>
								<ChunklistMaxAge>0</ChunklistMaxAge>
								<ChunklistWithDirectivesMaxAge>60</ChunklistWithDirectivesMaxAge>
								<SegmentMaxAge>-1</SegmentMaxAge>
								<PartialSegmentMaxAge>-1</PartialSegmentMaxAge>
							</CacheControl>
							
							<ChunkDuration>0.2</ChunkDuration>
							<PartHoldBack>3.0</PartHoldBack>
							<SegmentDuration>6</SegmentDuration>
							<SegmentCount>10</SegmentCount>
							<CrossDomains>
								<Url>*</Url>
							</CrossDomains>

						</LLHLS>

					</Publishers>

				</Application>

**Reply by getroot** [reply]
Here's a tip: When configuring a playlist, place the lowest bitrate at the top and organize it in bitrate order. This appears to be a bug in Safari. This is probably resolved in 17.2. (Currently in beta.)
